File: src/lib/calendar/resolveDay.ts
import type { DayStatus, SchoolLevel, SchoolYear, WeekPattern } from './types'

function dayOfWeek(iso: string): number {
  // ISO date strings are parsed as UTC midnight, so getUTCDay is stable
  return new Date(`${iso}T00:00:00Z`).getUTCDay() // 0 = Sunday, 6 = Saturday
}

export function resolveDay(
  year: SchoolYear,
  level: SchoolLevel,
  pattern: WeekPattern,
  // Contractually an ISO YYYY-MM-DD string produced by toIsoDate. Validated by
  // the caller (see validateSchoolYear.ts), not by this function -- resolveDay
  // does plain ISO string comparison and does not itself check calendar validity.
  date: string,
): DayStatus {
  const term = year.terms[level]
  if (date < term.start || date > term.end) {
    return { date, open: false, shortDay: false, reason: 'מחוץ לשנת הלימודים' }
  }

  const dow = dayOfWeek(date)
  if (dow === 6) return { date, open: false, shortDay: false, reason: 'שבת' }
  if (dow === 5 && pattern === 'five_day') {
    return { date, open: false, shortDay: false, reason: 'יום שישי' }
  }

  const closure = year.closures.find(
    (c) => c.levels.includes(level) && date >= c.from && date <= c.to,
  )
  if (closure) return { date, open: false, shortDay: false, reason: closure.name }

  const short = year.shortDays.find((s) => s.levels.includes(level) && s.date === date)
  if (short) return { date, open: true, shortDay: true, reason: short.name }

  return { date, open: true, shortDay: false, reason: null }
}

File: ios/MacharCore/Sources/MacharCore/ResolveDay.swift
import Foundation

/// Resolves one date for one child. A direct port of `src/lib/calendar/resolveDay.ts`;
/// the two implementations must stay behaviourally identical, and
/// `ResolveDayTests` mirrors the TypeScript test cases one for one.
///
/// Precedence, highest first: outside the school year, Saturday, Friday on a
/// five-day week, a closure, a short day. Closure beats short day on a date clash.
public func resolveDay(
    year: SchoolYear,
    level: SchoolLevel,
    pattern: WeekPattern,
    date: String
) -> DayStatus {
    let term = year.term(for: level)

    if date < term.start || date > term.end {
        return DayStatus(date: date, open: false, shortDay: false, reason: "מחוץ לשנת הלימודים")
    }

    // An invalid date cannot be resolved by weekday; treat it as closed rather
    // than guessing. Callers validate with CivilDate.isValid before this point.
    guard let dow = CivilDate.dayOfWeek(date) else {
        return DayStatus(date: date, open: false, shortDay: false, reason: nil)
    }

    if dow == 6 {
        return DayStatus(date: date, open: false, shortDay: false, reason: "שבת")
    }
    if dow == 5, pattern == .fiveDay {
        return DayStatus(date: date, open: false, shortDay: false, reason: "יום שישי")
    }

    if let closure = year.closures.first(where: {
        $0.levels.contains(level) && date >= $0.from && date <= $0.to
    }) {
        return DayStatus(date: date, open: false, shortDay: false, reason: closure.name)
    }

    if let short = year.shortDays.first(where: {
        $0.levels.contains(level) && $0.date == date
    }) {
        return DayStatus(date: date, open: true, shortDay: true, reason: short.name)
    }

    return DayStatus(date: date, open: true, shortDay: false, reason: nil)
}
